src/backend/utils/misc/queryenvironment.c

```diff
@@ -978,8 +1111,469 @@ void ENRDropEntry(Oid id)
+					}
+					break;
+				}
+			case SequenceRelationId:
+				{
+					foreach(lc2, enr->md.uncommitted_cattups[ENR_CATTUP_SEQUENCE])
+					{
+						ENRUncommittedTuple othertup = (ENRUncommittedTuple) lfirst(lc2);
+						Form_pg_sequence other = (Form_pg_sequence) GETSTRUCT(othertup->tup);
+						if (((Form_pg_sequence) GETSTRUCT(tup))->seqrelid == other->seqrelid)
+							return othertup->optype == ENR_OP_DROP;
+					}
+					break;
+				}
+			default:
+				break;
+		}
+	}
+	return false;
+}
+
+/*
+ * ENRCommitChanges
+ *
+ * 1. Free uncommitted tuple data from each ENR, then mark it as committed.
+ * 2. Finish freeing ENRs that were dropped in this transaction.
+ */
+void
+ENRCommitChanges(QueryEnvironment *queryEnv)
+{
+	ListCell 	*lc = NULL;
+	MemoryContext oldcxt;
+
+	if (!queryEnv)
+		return;
+
+	oldcxt = MemoryContextSwitchTo(queryEnv->memctx);
+
+	foreach(lc, queryEnv->namedRelList)
+	{
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		if (enr->md.enrtype != ENR_TSQL_TEMP || enr->md.is_table_variable)
+			continue;
+
+		ENRDeleteUncommittedTupleData(InvalidSubTransactionId, enr);
+		enr->md.is_committed = true;
+	}
+
+	lc = NULL;
+
+	foreach(lc, queryEnv->dropped_namedRelList)
+	{
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		pfree(enr->md.name);
+		pfree(enr);
+	}
+
+	list_free(queryEnv->dropped_namedRelList);
+	queryEnv->dropped_namedRelList = NIL;
+
+	MemoryContextSwitchTo(oldcxt);
+}
+
+/*
+ * ENRRollbackChanges
+ * 
+ * NB: This will also be called when an error occurs for transaction rollback. 
+ * 1. Move any ENRs that were dropped back into the main ENR list.
+ * 2. Iterate through the ENR list and for each ENR, go through uncommitted tuples
+ *    and perform the opposite operation to restore it.
+ * 3. Finally, clean up the uncommitted tuple list and perform all the necessary deletions
+ *    last to avoid potential for any dependency issues.
+ */
+void
+ENRRollbackChanges(QueryEnvironment *queryEnv)
+{
+	ListCell 	*lc;
+	List *enrs_to_drop = NIL;
+	MemoryContext oldcxt;
+
+	if (!queryEnv)
+		return;
+
+	oldcxt = MemoryContextSwitchTo(queryEnv->memctx);
+
+	/* Add back any ENRs that must be restored. */
+	foreach (lc, queryEnv->dropped_namedRelList)
+	{
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		Assert(enr->md.enrtype == ENR_TSQL_TEMP);
+		Assert(!enr->md.is_table_variable);
+
+		queryEnv->namedRelList = lappend(queryEnv->namedRelList, enr);	
+	}
+	list_free(queryEnv->dropped_namedRelList);
+	queryEnv->dropped_namedRelList = NIL;
+
+	/*
+	 * Loop through the registered ENRs to process each.
+	 */
+	foreach(lc, queryEnv->namedRelList)
+	{
+		Relation rel;
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		if (enr->md.enrtype != ENR_TSQL_TEMP || enr->md.is_table_variable)
+			continue;
+
+		if (!enr->md.is_committed)
+		{
+			enrs_to_drop = lappend(enrs_to_drop, enr);
+			continue;
+		}
+
+		/* ROLLBACK block. */
```

Comment:
what if there are multiple table versions all committed? like
begin;
create; drop; create; drop; create; drop; create;
commit
why are we keeping all versions and not just one?

Reply:
in above example, we will have one entry in named rel list and 3 entries in dropped named list. We are transferring all dropped entries into named list. Why are we doing that? we just need the first one from the list which is committed, others can be dropped immediately

Reply:
ack, added a fix in ENRDropEntry